feat: Handling git credentials via credential helper

hferentschik opened the pull request:
The key feature of this pull request is to make `jx` a git credential helper which allows for using our Kube based authentication for retrieving git push/pull/fetch credentials. This is a replacement for adding the password to the remote URL directly or the workaround of generating an authenticated git URL on the fly (which as side effect does not update the remote refs).

Some other changes deal with the fact to allow consecutive execution of `jx boot`, however, this is a temporary solution and does not address the core issues with `jx boot`.  

Note: Now it is possible to have multiple consecutive boot executions and regular commits made on master will be cherry-picked on local `jx boot` execution. This does not change the fact that we should not be using cherry-picking in the first place since it cannot handle merge commits. 

fixes #5772

hferentschik commented:
@ccojocar, @abayer - any chance I  can get a review?

abayer commented:
Seems fine to me, but I'd really like @ccojocar to review it.

ccojocar commented on pkg/gits/credentialhelper/git_credentialhelper.go:
```diff
@@ -0,0 +1,111 @@
+		err = h.Get()
+	case "store":
+		// not yet implemented (HF)
+		fmt.Println("")
```
Shouldn't we return an error here when is not implemented?

hferentschik replied:
I think in this case returning the empty string is the right thing. There is no need for an error. Git will send us the store request and we just ignore it.

Eventually, I think we could use this to store the user's credentials within the cluster/vault.

ccojocar commented on pkg/gits/credentialhelper/git_credentialhelper.go:
```diff
@@ -0,0 +1,111 @@
+		fmt.Println("")
+	case "erase":
+		// not yet implemented (HF)
+		fmt.Println("")
```
ditto

ccojocar commented on pkg/gits/credentialhelper/git_credentialhelper.go:
```diff
@@ -0,0 +1,111 @@
+		// not yet implemented (HF)
+		fmt.Println("")
+	default:
+		err = errors.Errorf("Invalid git credential operation '%s'", op)
```
error message shouldn't be capitalised

abayer commented:
Yeah, the local case is critical, but that all goes through import's `Run`, which calls `GitLocalAuthConfigService()`, so...it's probably fine?

ccojocar commented:
>Yeah, the local case is critical, but that all goes through import's Run, which calls GitLocalAuthConfigService(), so...it's probably fine?

I think the credentials retrieved through LocalAuthConfigService are used only for Git provider but not for git CLI. In which case, git will use what's configured into the system. This means this  credentials helper (if configured) which will fetch the credentials as well from the cluster.

abayer commented:
Good point.

hferentschik commented:
> I am also thinking about the use case when the git needs to use local credentials instead of the credentials configured into the cluster. For instance, when the jx create quickstart or import commands are executed locally. Probably in those cases, we should use the local credentials (credentials store) instead.

That's the beauty here. The git credential helpers are scoped. There is local (just per current repo), global (for all repos of the user) and system-wide (per git installation). Usually, on macOS for example, there is a system-wide credential helper configured which hooks into the keychain. Hence, I am doing:

```
func configureJxAsGitCredentialHelper(dir string, gitter Gitter) error {
	// configure jx as git credential helper for this repo
	jxProcessBinary, err := os.Executable()
	if err != nil {
		return errors.Wrapf(err, "unable to determine jx binary location")
	}
	return gitter.Config(dir, "--local", "credential.helper", fmt.Sprintf("%s step git credentials --credential-helper", jxProcessBinary))
}
```

In the case of the boot repo I configure the helper on the --local level. So it applies just for this repo. We are basically in control where and when we want jx as credential helper.

Also there is a nice chain, if git does not get a response from the local git credential helper, it asks the global one and finally the system wide one (provided any of them is set). This means for the case were we don't return credentials, git automatically continues trying to resolve the password.

Does that make sense?

hferentschik commented:
> Also I am not sure which are the implications from security perspective, e.g. if someone manages to trigger this command in a pipeline context and exfiltrate the git credentials configured into the cluster through the standard output of the command.

I guess that's somehow possible, but in getting in-/out-put redirection right is probably not so easy. Probably easier to just run `jx step git credentials` and then echo the generated file. Seems much easier. I guess that's where sanitizing would come into play. Not sure.

hferentschik commented on pkg/cmd/step/git/credentials/step_git_credentials.go:
```diff
@@ -108,13 +107,12 @@ func (o *StepGitCredentialsOptions) Run() error {
-			user:       string(secret.Data["user"]),
-			password:   string(secret.Data["token"]),
-			serviceURL: string(secret.Data["url"]),
+		creds, err := credentialhelper.CreateGitCredentialFromURL(string(secret.Data["url"]), string(secret.Data["token"]), string(secret.Data["user"]))
```
You might be talking to the wrong person. @garethjevans  introduced this in e9aa7f3840da37c6e706856785e2b1759867b749. I just rebased on top of this change here and made it work with the changes I am making. I actually meant to ask you whether you have seen this change since I was also wondering why we would be reading directly from the secret. @garethjevans might have an answer to this.

@ccojocar is there an easy fix to this? Do you want me to address this as part of this PR?

ccojocar replied:
If we want to have this working in all setup combinations, we need to switch to `AuthConfigService`. Actually, there is a `GitAuthConfigService` in the `CommonOptions` which loads the auth config and abstracts all the secrets lookup based on the secret source and the setup available. In the latest boot version, that secret is not anymore present. There is a config map with URI references into the vault.

hferentschik replied:
What I am confused about is that we are using already a AuthConfigService, at least for the code which existed. I don't understand why we are reading this secret directly in this case.

ccojocar replied:
Looking closer to that code and it seems that the loading directly from secrets is activated only when there is some value in the `JX_CREDENTIALS_FROM_SECRET`.  I think we can ignore my comment for now.

hferentschik replied:
works for me :-)

hferentschik commented on pkg/cmd/step/git/credentials/step_git_credentials.go:
```diff
@@ -144,19 +142,38 @@ func (o *StepGitCredentialsOptions) Run() error {
+		// the credential helper operation (get|store|remove) is passed as last argument to the helper
+		err = helper.Run(os.Args[len(os.Args)-1])
+		if err != nil {
+			return errors.Wrap(err, "error responding to git credential request")
```
OOI, how do you know that? How do you know that for this error no wrapping is needed? Is this something you know from experience or is there some sort of rule behind it?

hferentschik replied:
Or it's because we are still in _Run_, right?

hferentschik commented on pkg/cmd/step/verify/step_verify_environments.go:
```diff
@@ -480,12 +480,19 @@ func (o *StepVerifyEnvironmentsOptions) pushDevEnvironmentUpdates(environmentRep
+		return errors.Wrapf(err, "no remote configured for %s", environmentRepo.CloneURL)
+	}
+
+	err = gitter.Push(localRepoDir, remoteName, true, "master")
```
yes

hferentschik commented on pkg/gits/credentialhelper/git_credential.go:
```diff
@@ -0,0 +1,122 @@
+func CreateGitCredential(lines []string) (GitCredential, error) {
+	var credential GitCredential
+
+	if lines == nil {
```
well, technically this will just mean that there won't be anything to extract. _nil_ on the other hand could panic.

hferentschik replied:
Instead of checking the lines length it would maybe be more useful to check whether there is at least a Protocol and Host at the end. I think for now we can live with how it is.

hferentschik commented on pkg/gits/credentialhelper/git_credential.go:
```diff
@@ -0,0 +1,122 @@
+		return credential, errors.Wrap(err, "unable to extract git credential parameters")
+	}
+	for key, value := range fieldMap {
+		v := reflect.ValueOf(&credential).Elem().FieldByName(strings.Title(key))
```
> Why do you need this trick? I am not a bit fan of reflect in production code even though makes the code more generic.

I don't need it :-) I think I just did not want to type things out. As you say, it's a bit more generic. Any concrete reason why you don't like reflection? To be fair, I was undecided on whether I should type things our or whether I should use reflection. Coming from the Java world I am quite used to using reflection, but it might not be as common in Go. 

If you feel strongly about it, I can change it

hferentschik replied:
> I think that you can also use json.Marshal(fieldMap) and then json.Unmarshal(&GitCredentials). It is a bit more portable if the types of the struct gets changed.

that's probably one way, even though this under the hood probably does the same type of reflection as well. Really, if it comes down to it, you can just type it out. 

To be fair, people might be more familiar with the Marshalling technique.

hferentschik replied:
Actually there is a difference, your marshalling technique will silently ignore unknown fields.

hferentschik replied:
Using the JSON technique since it is more familiar

abayer commented:
When @ccojocar and I are talking about "local" here, we're talking about: https://github.com/jenkins-x/jx/blob/8a527746ce43826f06ad41da4902920450930a7a/pkg/cmd/importcmd/import.go#L244 That forces usage of `~/.jx/gitAuth.yaml` rather than any credentials in the cluster. This is used in `jx import` and `jx create quickstart` (and the legacy `jx create spring`) but nowhere else, so far as I'm aware.

hferentschik commented:
> Yeah, the local case is critical, but that all goes through import's Run, which calls GitLocalAuthConfigService(), so...it's probably fine?

Most importantly, the git credentials helpers are chained as explained earlier. What we are missing in general and which we should address asap is proper user management in terms of Jenkins X. We need a distinction between different types of users (dev vs admin). I think to truly sort the git integration, we will need to address this since imo they go hand in hand.

hferentschik commented:
@abayer, @ccojocar - thanks for the review and discussion :-) I rebased the PR and addressed the suggested changes which imo made sense. 

There is the outstanding question around reading directly the Kubernetes secret, but as said, I did not introduce it. Not sure what you want me to do with it.

Any chance you have tried this change locally and tried boot and maybe even a reboot? Would be nice to get some more sanity checks.

ccojocar commented:
> What we are missing in general and which we should address asap is proper user management in terms of Jenkins X. We need a distinction between different types of users (dev vs admin). I think to truly sort the git integration, we will need to address this since imo they go hand in hand.

There is an issue related to this and it is pretty high on my security list https://github.com/jenkins-x/jx/issues/6453. We need also to think about deprecating `teams` and `users` CRDs in favour of standard user authentication.

> There is the outstanding question around reading directly the Kubernetes secret, but as said, I did not introduce it. Not sure what you want me to do with it.

This we need to fix if we want to have this working in all setup combination. Since the introduction of `jx-auth-config` ConfigMap in combination with vault, that secret is not anymore available. See my comment above. 

> That forces usage of ~/.jx/gitAuth.yaml rather than any credentials in the cluster. This is used in jx import and jx create quickstart (and the legacy jx create spring) but nowhere else, so far as I'm aware.

Please can you try to test it with `create quickstart` or `import` to check that we still use the local git credentials instead of the pipeline credentials installed into the cluster when pushing the local git repository created from a quickstart.  

It is not clear to me, is now the `configureJxAsGitCredentialHelper` always called to setup the credentials helper for git CLI wrapper, or only specifically in boot/verify environment?

hferentschik commented:
> Please can you try to test it with create quickstart or import to check that we still use the local git credentials instead of the pipeline credentials installed into the cluster when pushing the local git repository created from a quickstart.

Ok will do

> It is not clear to me, is now the configureJxAsGitCredentialHelper always called to setup the credentials helper for git CLI wrapper, or only specifically in boot/verify environment?

well, everywhere where we call `ForkAndPullRepo` really. That is mainly boot/verify environment, but there is also used in `CreatePullRequest`. That said, it does not change the fact that in its current form in all these cases the pipeline users is used. This is not something which is changed by this PR.

ccojocar commented:
@hferentschik Is this also fixing https://github.com/jenkins-x/jx/issues/5463?

hferentschik commented:
@ccojocar - double-checked with import and quickstart and it works

ccojocar commented:
/lgtm

hferentschik commented:
>  Is this also fixing #5463?

It is putting the groundwork in place for it. It is not integrated with import though. The idea would be to store the token provided by the user in the cluster/vault (opposed to adding it to the remote URL or adding it to the local gitAuth.yaml). Then we would use the credential helper mechanism to retrieve the token and authenticate. 

Is there a reason we don't already store the token the user provides in Vault?

ccojocar commented:
> Is there a reason we don't already store the token the user provides in Vault?

We can do this, but right now we don't have security policies defined per users in vault, also the user authentication is not enabled, only the kubernetes authentication. In this case everything in one bucket. See https://github.com/jenkins-x/jx/issues/6605.
